### ISSUE TITLE
Running AI training tasks in kata containers is much slower than runc

### ISSUE THREAD

**Issue by soulseen** [kata-containers/kata-containers#5833]
# Description of problem

I test an AI training task in runc and kata respectively, but it was significantly slower in kata.
I analysed the profiles in both cases and found that the difference was that kata was significantly less efficient than runc when saving the model objects to a file.

**kata save model cost:**
![image](https://user-images.githubusercontent.com/16031013/205573556-7f69fe10-ad9d-444d-baa5-8fa47f397d20.png)

**runc save model cost:**
![image](https://user-images.githubusercontent.com/16031013/205573601-fc31dfd5-7fac-4950-be8f-ef30edb0837d.png)

After in-depth investigation, I found that the root cause of kata's inefficiency was that copying model objects from GPU memory to CPU memory took about twice as long as `runc`:
the model size is about 9G.
`kata` copy model from GPU memory to CPU memory cost: 12.351327s
`runc` copy model from GPU memory to CPU memory cost: 5.296830s

# Expected result

speed up copy model from GPU memory to CPU memory in `kata`, At least close to the speed of `runc`.

# Actual result

(replace this text with an explanation of what actually happened)

# Further information

(replace this text with any extra information you think might be useful)

**Comment by soulseen** [comment]
@liubin @zvonkok can help to take a look? thanks.

**Comment by soulseen** [comment]
How to make the GPU to CPU memory copy faster in kata containers?

**Comment by sameo** [comment]
Hey @soulseen, could you describe what you mean here by copying a model from the GPU memory to CPU memory? I assume you're using a specific toolkit that eventually triggers DMA from the GPU address space into the CPU physical memory?
Also, you seem to imply that the model is then saved to a file. Do we know how long the actual storage time takes in Kata vs runc?
Last question: Which GPU are you using to run this?

**Comment by soulseen** [comment]
@sameo yes, and I used [_save](https://github.com/pytorch/pytorch/blob/v1.7.1/torch/serialization.py#L491) function in PyTorch to save the model to a file, and as you said, it first triggers DMA from the GPU address space into the CPU physical memory, then save to a file.

> Do we know how long the actual storage time takes in Kata vs runc?

sure, this is the sample result:
kata storage time: 1.515632
runc storage time: 1.507609
I think the storage time they are very close.

>  Last question: Which GPU are you using to run this?

GPU card: NVIDIA GeForce GTX 1080 Ti

**Comment by soulseen** [comment]
> Hey @soulseen, could you describe what you mean here by copying a model from the GPU memory to CPU memory? I assume you're using a specific toolkit that eventually triggers DMA from the GPU address space into the CPU physical memory? 

hi, @sameo After verification, I don't think it uses DMA to transfer GPU memory to CPU memory.
here is the DMA `bytes_transferred`
```
$ cat /sys/class/dma/dma*/bytes*
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
```

**Comment by soulseen** [comment]
maybe this patch can resolve this issue.
https://github.com/qemu/qemu/commit/c2a490e344b4e231cf9488c67df7ee46977b1ebe
but kata did not expose the PCIe link speed and width to users and defaulted only at 2.5GT/s & x1 lanes.
@sameo @liubin

**Comment by liubin** [comment]
@soulseen this [commit](https://github.com/qemu/qemu/commit/a09d2038cc5f8e45e2126461e5fb1eb9f4874be3) has changed the default values to the max value, so this may not solve your problem. And the [changelog](https://wiki.qemu.org/ChangeLog/4.0) also says the same default value.

**Comment by zvonkok** [comment]
@soulseen Taking a look.
